### PR TITLE
Feature/multiple storage classes

### PR DESCRIPTION
**Pull request by meln5674**
* Fixes https://github.com/rancher/local-path-provisioner/issues/38
* Fixes https://github.com/rancher/local-path-provisioner/issues/348

While is is currently possible to deploy multiple instances of the provisioner to support multiple storage classes, as well as to support both RWO, RWX, and ROX volumes at the same time, the additional resource consumption and deployment complexity are undesirable.

This patch adds an additional field to the `config.json` named `storageClassConfigs`, which is mutually exclusive with the existing `nodePathMap` and `sharedFilesystemPath` fields. This field is an object, whose keys are the names of storage classes, and its values are nested objects that contain either `nodePathMap` or `sharedFilesystemPath`. When a volume request is received, if this field is set, it is consulted for the configuration to use based on the name of the storage class, otherwise the original configuration fields are used.

This patch also updates the helm chart to add a matching option, which also deploys multiple storage class objects if selected.

This change is backwards compatible, installing this version of the provisioner should not break any existing installs.

One advantage of this change is that single-node k3s/k3d/kind clusters will be able to provide RWX and ROX storage classes out of the box.

### CODE REVIEW

**Comment by derekbit** [comment]
Can you help rebase the commit and update the README.md as well? Thank you.

**Comment by meln5674** [comment]
Rebased. Can you elaborate on what sort of update to the README you're looking for? I did update the config.json definition section to include the new field I added.

FYSA I will be away from home for the next week, so if you'd like to put this on hold until then, I don't mind.

**Comment by derekbit** [comment]
@meln5674 
Sorry, I missed the README.md is updated.
BTW, it would be nice if you would be able to add a test for the feature. Thank you.

**Comment by meln5674** [comment]
Sorry for the delay on this, I ended up falling seriously ill after getting back from my trip.

I can't seem to find any information on how the tests here are supposed to work.

My default guess of `make test` shows `[no test files]`. I found `scripts/test` to be running `go test -cover -tags=test ./...`, which produces the same result. I noticed in `test/pod_test.go` its only being built when the `e2e` tag is present, so I tried `go test -cover -tags=e2e ./...`, which at least did something, but that produced the error `ERROR: a list of image names is required` which also happens on the master branch.

I did in fact write tests while writing this, though since I ran into the above issues since the start, I just wrote them in a shell script, which I've put into [this gist](https://gist.github.com/meln5674/2e39647d15e2ed99addb54ead0150ce1). If you give me any pointers on how to actually run the tests, as well as any on how best to integrate what I've written into the existing test setup, I'd really appreciate it.

**Comment by timvahlbrock** [comment]
Hey @meln5674 , hope you are well again. I don't have any experience with Go and am still somewhat new to kubernetes, but as I'm looking forward to this feature I played around with the repo a bit. I checked how the CI is running the tests using the `.drone.yaml` file: It runs them using `dapper e2e-test`. I think dapper will be installed within the dir when executing the make file, so I ran `./.dapper e2e-test`. This actually runs a lot of tests (takes some time to get started though), but `TestPodWithRWOPVolume` is already failing on master in my case (even though CI is green). Hope this helps.

**Comment by meln5674** [comment]
@timvahlbrock Thanks for the detective work, and I am in fact in better health again, thank you for your concern.

@derekbit I think I've worked out how the tests are structured now, and I've added a test which creates two storage classes, two matching PVC's, and mounts them to a single pod to match how the other tests work so I wouldn't have to re-work how the tests are structured. Let me know if you'd prefer I go back and do a 2-pod test instead.

**Comment by timvahlbrock** [comment]
@meln5674 is this pr possibly not being reviewed because the pipeline is red (even though that's because it was killed...)?

**Comment by meln5674** [comment]
I saw there were a few other PR builds that were killed, each because they never started their s390x build, so I assumed the project was working through some CI/CD issues.

**Comment by timvahlbrock** [comment]
Can you set @derekbit as a reviewer explicitly or something? Or try to push something again to get the pipeline green? I don't wanna be too annoying or something, but other PRs seem to be getting merged.

**Comment by timvahlbrock** [comment]
@meln5674 do you have permission to request a review on this PR?

**Comment by derekbit** [comment]
> Can you set @derekbit as a reviewer explicitly or something? Or try to push something again to get the pipeline green? I don't wanna be too annoying or something, but other PRs seem to be getting merged.

@timvahlbrock Sorry for the late reply. Busy at Longhorn project.

> Or try to push something again to get the pipeline green

The error is due to that the s390x machine doesn't work.
amd64 and arm64 passed the tests. 
Are all updates ready to review? If yes, I will fix it and review the PR. soon.

**Comment by meln5674** [comment]
Yes, everything should be ready for review.

**Comment by timvahlbrock** [comment]
@derekbit  Slipped out of ToDo's again?

File: README.md
Comment (derekbit) [inline README.md:205]: IIUC, the backward compatibility of an existing configmap with `nodePathMap` is also addressed, right?

Comment (meln5674) [reply README.md:205]: Yes. The ConfigMap under this patch can contain one and only one of nodePathMap, sharedFileSystemPath, or storageClassConfigs. An existing ConfigMap with one of nodePathMap and sharedFileSystemPath should continue to operate as-is, which I believe the tests demonstrate.

File: provisioner.go
Comment (derekbit) [inline provisioner.go:260]: This handles he `nodePathMap` as well as address the existing configmap format.

File: test/testdata/multiple-storage-classes/local-path-config.yaml
Comment (derekbit) [inline test/testdata/multiple-storage-classes/local-path-config.yaml:22]: https://raw.githubusercontent.com/rancher/local-path-provisioner/v0.0.26/deploy/local-path-storage.yaml using 
```
  config.json: |-
    {
            "nodePathMap":[
            {
                    "node":"DEFAULT_PATH_FOR_NON_LISTED_NODES",
                    "paths":["/opt/local-path-provisioner"]
            }
            ]
    }
```

Do you suggest using
```
  config.json: |-
    {
            "storageClassConfigs":{
                   "local-path": {
                          "nodePathMap": [
                          {
                                  "node":"DEFAULT_PATH_FOR_NON_LISTED_NODES",
                                  "paths":["/opt/local-path-provisioner"]
                          }
                          ]
                   }
            }
    }
```
is better?

The we can deprecate the old confg.

**Review by derekbit — APPROVED** [review]
The implementation LGTM.
Only one question for the deploy manifest.

File: test/testdata/multiple-storage-classes/local-path-config.yaml
Comment (meln5674) [reply test/testdata/multiple-storage-classes/local-path-config.yaml:22]: Not sure. It might make sense to leave it as-is so that existing installations don't change unexpectedly (even if the behavior is identical), and using the new config would be left as opt-in.

**Comment by timvahlbrock** [comment]
Is there anything I can help with to progress this further?

**Comment by derekbit** [comment]
@timvahlbrock
The PR looks good to me and is included in the upcoming v0.0.27. I will merge it after the broken CI is fixed. Thank you. :)

**Review by derekbit — APPROVED** [review]

**Comment by JuniorJPDJ** [comment]
how nice, thanks! I'm waiting for this to be released!

**Comment by arjunkanti** [comment]
When is v0.0.27 version getting released with this change..?

> @timvahlbrock The PR looks good to me and is included in the upcoming v0.0.27. I will merge it after the broken CI is fixed. Thank you. :)

**Pull request merged**
